*************
Publications
*************

Here we provide links to all publications and preprints detailing our software and SARS-CoV-2 phylogenomics analyses.

.. toctree::
   :hidden:
   Publications.rst

- Turakhia Y, Thornlow B, Hinrichs A, De Maio N, Gozashti L, Lanfear R, Haussler D, and Corbett-Detig R. `**Ultrafast Sample Placement on Existing Trees (UShER) Empowers Real-Time Phylogenetics for the SARS-CoV-2 Pandemic.**, <https://www.biorxiv.org/content/10.1101/2020.09.26.314971v1.abstract>`_ Nature Genetics. In Press.

- Turakhia Y, De Maio N, Thornlow B, Gozashti L, Lanfear R, Walker C, Hinrichs A, Fernandes J, Borges R, Slodkowicz G, Weilguny L, Haussler D, Goldman N, and Corbett-Detig R. `**Stability of SARS-CoV-2 Phylogenies.**, <https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1009175>`_ PLOS Genetics. 2020. 16(11): e1009175.

- De Maio N, Walker C, Turakhia Y, Lanfear R, Corbett-Detig R, and Goldman N. `**Mutation rates and selection on synonymous mutations in SARS-CoV-2.**, <https://www.biorxiv.org/content/10.1101/2021.01.14.426705v1.abstract>`_ bioRxiv. 2020.

- DeMaio N, Walker C, Borges R, Weilguny L, Slodkowicz G, and Goldman N. `**Issues with SARS-CoV-2 sequencing data.**, <http://virological.org/t/issues-with-sars-cov-2-sequencing-data/473>`_ Virological. 2020.

- Gozashti L, Walker C, Goldman N, Corbett-Detig R, and DeMaio N. `**Updated analysis with data from 13th November 2020.**, <https://virological.org/t/issues-with-sars-cov-2-sequencing-data/473/14>`_ Virological. 2020.


